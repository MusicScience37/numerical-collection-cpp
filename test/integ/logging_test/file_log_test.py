"""Test of file logs."""

import pathlib
import subprocess

import approvaltests

from .create_scrubber import create_scrubber


def _verify_command_result(
    writer_path: pathlib.Path,
    config_file_path: pathlib.Path,
    test_temp_dir_path: pathlib.Path,
    log_path: pathlib.Path,
) -> None:
    result = subprocess.run(
        args=[str(writer_path), str(config_file_path)],
        cwd=str(test_temp_dir_path),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        check=False,
        encoding="utf-8",
    )

    with open(str(log_path), mode="r", encoding="utf-8") as file:
        log_contents = file.read()

    approvaltests.verify(
        f"""exit code: {result.returncode}
stdout:
{result.stdout}
stderr:
{result.stderr}
{log_path.name}:
{log_contents}
""",
        options=approvaltests.Options().with_scrubber(
            create_scrubber(str(config_file_path))
        ),
    )


def test_trace_log(
    writer_path: pathlib.Path,
    test_temp_dir_path: pathlib.Path,
    config_dir_path: pathlib.Path,
):
    """Test with trace log level."""

    _verify_command_result(
        writer_path=writer_path,
        config_file_path=config_dir_path / "file_log_trace.toml",
        test_temp_dir_path=test_temp_dir_path,
        log_path=test_temp_dir_path / "test_integ_logging.log",
    )


def test_combined_file_console_log(
    writer_path: pathlib.Path,
    test_temp_dir_path: pathlib.Path,
    config_dir_path: pathlib.Path,
):
    """Test of combined file and console logs."""

    _verify_command_result(
        writer_path=writer_path,
        config_file_path=config_dir_path / "combined_file_console_log.toml",
        test_temp_dir_path=test_temp_dir_path,
        log_path=test_temp_dir_path / "test_integ_logging.log",
    )


def test_async_file_log(
    writer_path: pathlib.Path,
    test_temp_dir_path: pathlib.Path,
    config_dir_path: pathlib.Path,
):
    """Test of asynchronous file logs."""

    _verify_command_result(
        writer_path=writer_path,
        config_file_path=config_dir_path / "async_file_log.toml",
        test_temp_dir_path=test_temp_dir_path,
        log_path=test_temp_dir_path / "test_integ_logging.log",
    )


def test_async_file_console_log(
    writer_path: pathlib.Path,
    test_temp_dir_path: pathlib.Path,
    config_dir_path: pathlib.Path,
):
    """Test of asynchronous logs to a file and console."""

    _verify_command_result(
        writer_path=writer_path,
        config_file_path=config_dir_path / "async_file_console_log.toml",
        test_temp_dir_path=test_temp_dir_path,
        log_path=test_temp_dir_path / "test_integ_logging.log",
    )
