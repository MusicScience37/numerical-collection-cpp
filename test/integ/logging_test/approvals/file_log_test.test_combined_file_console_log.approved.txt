exit code: 0
stdout:
[<time-stamp>] [warning] [example_tag] warning (writer.cpp:49, write_logs)
[<time-stamp>] [error] [example_tag] error (writer.cpp:50, write_logs)
[<time-stamp>] [critical] [example_tag] critical (writer.cpp:51, write_logs)
[<time-stamp>] [warning] [] warning (writer.cpp:65, write_to_default_tag)
[<time-stamp>] [error] [] error (writer.cpp:66, write_to_default_tag)
[<time-stamp>] [critical] [] critical (writer.cpp:67, write_to_default_tag)
[<time-stamp>] [warning] [example_tag] warning (writer.cpp:81, write_logs_with_macros)
[<time-stamp>] [error] [example_tag] error (writer.cpp:82, write_logs_with_macros)
[<time-stamp>] [critical] [example_tag] critical (writer.cpp:83, write_logs_with_macros)

stderr:

num_collect_test_integ_logging.log:
[<time-stamp>] [debug] [example_tag] debug (writer.cpp:44:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag] iteration (writer.cpp:45:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag] iteration_label (writer.cpp:46:<col>, <function-name>)
[<time-stamp>] [summary] [example_tag] summary (writer.cpp:47:<col>, <function-name>)
[<time-stamp>] [info] [example_tag] info (writer.cpp:48:<col>, <function-name>)
[<time-stamp>] [warning] [example_tag] warning (writer.cpp:49:<col>, <function-name>)
[<time-stamp>] [error] [example_tag] error (writer.cpp:50:<col>, <function-name>)
[<time-stamp>] [critical] [example_tag] critical (writer.cpp:51:<col>, <function-name>)
[<time-stamp>] [debug] [] debug (writer.cpp:60:<col>, <function-name>)
[<time-stamp>] [iteration] [] iteration (writer.cpp:61:<col>, <function-name>)
[<time-stamp>] [iteration] [] iteration_label (writer.cpp:62:<col>, <function-name>)
[<time-stamp>] [summary] [] summary (writer.cpp:63:<col>, <function-name>)
[<time-stamp>] [info] [] info (writer.cpp:64:<col>, <function-name>)
[<time-stamp>] [warning] [] warning (writer.cpp:65:<col>, <function-name>)
[<time-stamp>] [error] [] error (writer.cpp:66:<col>, <function-name>)
[<time-stamp>] [critical] [] critical (writer.cpp:67:<col>, <function-name>)
[<time-stamp>] [debug] [example_tag] debug (writer.cpp:76:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag] iteration (writer.cpp:77:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag] iteration_label (writer.cpp:78:<col>, <function-name>)
[<time-stamp>] [summary] [example_tag] summary (writer.cpp:79:<col>, <function-name>)
[<time-stamp>] [info] [example_tag] info (writer.cpp:80:<col>, <function-name>)
[<time-stamp>] [warning] [example_tag] warning (writer.cpp:81:<col>, <function-name>)
[<time-stamp>] [error] [example_tag] error (writer.cpp:82:<col>, <function-name>)
[<time-stamp>] [critical] [example_tag] critical (writer.cpp:83:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]         val1        val2        val3 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]            0         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]           10         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]           20         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]           30         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]           40         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]           50         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]           60         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]           70         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]           80         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]           90         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]          100         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]          110         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [iteration] [example_tag]          120         abc       1.235 (writer.cpp:110:<col>, <function-name>)
[<time-stamp>] [summary] [example_tag] Finished iterations: val1=122, val2=abc, val3=1.235 (writer.cpp:114:<col>, <function-name>)
[<time-stamp>] [debug] [example_tag] Write a parameter: 12345. (writer.cpp:122:<col>, <function-name>)
[<time-stamp>] [info] [example_tag] Write multiple parameters: [1.234,5.678], abc (writer.cpp:126:<col>, <function-name>)

